### ISSUE TITLE
[rush] exclude certain files from `rush change` diff calculation

### ISSUE THREAD

**Issue by mmkal** [microsoft/rushstack#2263]
**Is this a feature or a bug?**

- [x] Feature
- [ ] Bug

**Please describe the actual behavior.**

If a project edits a test file, and no production code, `rush change` demands that a changefile be generated even though there are no (meaningful) updates required.

**What is the expected behavior?**

Allow configuring files to ignore from the diff calculation. One option could be `.npmignore`, or to use the package.json `files` property. This may need to follow source maps to avoid false negatives for packages that don't include "source" typescript, though. It may be less complex to pass in directly, e.g. `rush change ... --ignore '**/*.test.ts' --ignore '**/*.md'`, or to put something like `"ignoreChanges": ["**/*.test.ts", "**/*.md"]` in rush.json.

**If this is a bug, please provide the tool version, Node.js version, and OS.**

* **Tool:** rush
* **Tool Version:** 5.34.3

**Comment by octogonz** [comment]
This is a great idea. My recommendation is to use `.npmignore` like `rush deploy` does. 

We try to avoid putting settings like `"files"` in package.json because (1) that file lacks a strict spec and thus cannot be validated and (2) due to the design of the NPM protocol, the package.json file size impacts installation times.

**Comment by mmkal** [comment]
Fair enough - I agree that `.npmignore` is usually a better option that `files`. One issue with using it for this feature though, copied from the gitter thread:

What if I have `src` in my `.npmignore`, because I only publish `lib`? Then almost all changes would be excluded, unless rush can predict how `src/foo.test.ts` relates to `lib/foo.test.js` - which isn't trivial without building first.

**Comment by octogonz** [comment]
We do this: https://github.com/microsoft/rushstack/blob/master/libraries/node-core-library/.npmignore

**Comment by mmkal** [comment]
Right - looks like for that library, if I edited something in `src`, (say, [src/Constants.ts](https://github.com/microsoft/rushstack/blob/master/libraries/node-core-library/src/Constants.ts)), it'd be excluded by the `.npmignore`. So with this proposal there'd be no changefile generated?

**Comment by octogonz** [comment]
> One option could be `.npmignore`, or to use the package.json `files` property. This may need to follow source maps to avoid false negatives for packages that don't include "source" typescript, though. 

Right, hmm... I didn't realize that.  Source maps are not going to be reliable enough.

> It may be less complex to pass in directly, e.g. `rush change ... --ignore '**/*.test.ts' --ignore '**/*.md'`

This is not a good user experience. The most important audience for `rush change` are people who are totally unfamiliar with how the system works.

> or to put something like `"ignoreChanges": ["**/*.test.ts", "**/*.md"]` in rush.json.

Yeah.  It feels like it should be designed as an ignore file like `.gitignore` and `.npmignore`, since individual projects may have special requirements.  (`.changeignore`?)

Maybe we could make these files [riggable](https://rushstack.io/pages/heft/rig_packages/)?

**Comment by octogonz** [comment]
CC @iclanton

**Comment by iclanton** [comment]
I like the idea of `.changeignore` with the same syntax and behavior as `.gitignore` and `.npmignore`.

I think making this riggable will be more confusing that useful.

@mmkal - is this something you'd be interested in implementing?

**Comment by mmkal** [comment]
@iclanton opened a PR! #2279

**Comment by dmichon-msft** [comment]
I am of the opinion that the default recursive nesting behavior of .gitignore and .npmignore gets really confusing as well (it's also slow, since it requires a bunch of otherwise unnecessary file system operations) on every transaction. I'd prefer to have a policy of there being exactly one file in each project root, though I guess we can evaluate performance.
